confdb/confdb_control.go

```diff
@@ -31,134 +31,312 @@ var (
+		}
 	}
+
+	return clone
 }
 
-// convertToAuthenticationMethods converts []string to []AuthenticationMethod and validates it.
-func convertToAuthenticationMethods(methods []string) ([]AuthenticationMethod, error) {
-	sort.Slice(methods, func(i, j int) bool {
-		return methods[i] < methods[j]
-	})
+// authentication limits what keys can be used to sign messages used to remotely manage confdbs.
+type authentication uint8
 
-	// remove duplicates
-	methods = unique(methods)
+const (
+	// Only the operator's keys can be used to sign the messages.
+	OperatorKey authentication = 1 << iota
+	// Messages can be signed on behalf of the operator by the store.
+	Store
+)
 
-	var result []AuthenticationMethod
+const (
+	allAuth authentication = OperatorKey | Store
+)
+
+// newAuthentication converts []string to authentication and validates it.
+func newAuthentication(methods []string) (authentication, error) {
+	var auth authentication
 	for _, method := range methods {
-		if !isValidAuthenticationMethod(method) {
-			return nil, fmt.Errorf("invalid authentication method: %s", method)
+		switch method {
+		case "operator-key":
+			auth |= OperatorKey
+		case "store":
+			auth |= Store
+		default:
+			return 0, fmt.Errorf("invalid authentication method: %s", method)
 		}
-		result = append(result, AuthenticationMethod(method))
 	}
-	return result, nil
+	return auth, nil
 }
 
-// Operator holds the delegations for a single operator.
-type Operator struct {
-	ID     string
-	Groups []*ControlGroup
+// Convert authentication to a sorted []string.
+func (a authentication) toStrings() []string {
+	var keys []string
+	if a&OperatorKey == OperatorKey {
+		keys = append(keys, "operator-key")
+	}
+
+	if a&Store == Store {
+		keys = append(keys, "store")
+	}
+
+	return keys
 }
 
-// ControlGroup holds a set of views delegated through the given authentication.
-type ControlGroup struct {
-	Authentication []AuthenticationMethod
-	Views          []*ViewRef
+// operator holds the delegations for a single operator.
+type operator struct {
+	ID          string
+	Delegations map[viewRef]authentication
 }
 
-// ViewRef holds the reference to account/confdb/view as parsed from the
+// viewRef holds the reference to account/confdb/view as parsed from the
 // confdb-control assertion.
-type ViewRef struct {
+type viewRef struct {
 	Account string
 	Confdb  string
 	View    string
 }
 
-// AddControlGroup adds the group to an operator under the given authentication.
-func (op *Operator) AddControlGroup(views, auth []string) error {
-	if len(auth) == 0 {
-		return errors.New(`cannot add group: "auth" must be a non-empty list`)
-	}
-
-	authentication, err := convertToAuthenticationMethods(auth)
-	if err != nil {
-		return fmt.Errorf("cannot add group: %w", err)
-	}
-
-	if len(views) == 0 {
-		return errors.New(`cannot add group: "views" must be a non-empty list`)
-	}
+// String returns the string representation of the viewRef.
+func (v *viewRef) string() string {
```

Comment:
not sure if the lowercase "string()" is intentional since the comment is uppercase. I'd opt to use `String` so this can be used with things that expect `Stringer`